File: src/main/java/org/activecraft/activecraftcore/guicreator/GuiListener.java
package org.activecraft.activecraftcore.guicreator;

import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryCloseEvent;
import org.bukkit.inventory.InventoryView;
import org.bukkit.inventory.ItemStack;

public class GuiListener implements Listener {

    @EventHandler
    public void onInvClose(InventoryCloseEvent event) {
        GuiCreator guiCreator = GuiCreator.ofInventory(event.getInventory());
        if (guiCreator == null) return;

        if (guiCreator.getIdentifier().startsWith(GuiCreatorDefaults.CONFIRMATION_PREFIX)) {
            GuiCancelEvent guiConfirmEvent = new GuiCancelEvent(event.getInventory(), guiCreator.getIdentifier().replace(GuiCreatorDefaults.CONFIRMATION_PREFIX, ""),
                    (Player) event.getPlayer());
            Bukkit.getPluginManager().callEvent(guiConfirmEvent);
        }
    }

    @EventHandler
    public void onItemClick(InventoryClickEvent event) {
        if (event.getCurrentItem() == null || event.getCurrentItem().getType() == Material.AIR) return;
        ItemStack itemStack = event.getCurrentItem();
        Player player = (Player) event.getWhoClicked();
        Gui gui = GuiNavigator.getActiveGui(player);
        if (gui == null) return;
        if (gui.getInventory() != event.getClickedInventory()) return;
        GuiItem guiItem = gui.getGuiCreator().getCorrespondingGuiItem().get(itemStack);
        if (guiItem == null) return;

        //call GuiClickEvent
        GuiClickEvent guiClickEvent = new GuiClickEvent(guiItem, event);
        Bukkit.getPluginManager().callEvent(guiClickEvent);
        if (guiClickEvent.isCancelled()) event.setCancelled(true);
    }

    @EventHandler
    public void onGuiClick(GuiClickEvent event) {
        GuiItem item = event.getCurrentItem();
        InventoryView view = event.getView();
        Player player = (Player) view.getPlayer();
        if (item.getClickSound() != null)
            player.playSound(player.getLocation(), item.getClickSound(), 1f, 1f);
        item.getClickListenerList().forEach(guiAction -> guiAction.onClick(event));
        event.setCancelled(!item.isMovable());
    }
}

File: src/main/java/org/activecraft/activecraftcore/guicreator/GuiListener.kt
package org.activecraft.activecraftcore.guicreator

import org.activecraft.activecraftcore.guicreator.GuiCreator.Companion.ofInventory
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.inventory.InventoryCloseEvent

class GuiListener : Listener {
    @EventHandler
    fun onInvClose(event: InventoryCloseEvent) {
        val guiCreator = ofInventory(event.inventory) ?: return
        if (guiCreator.identifier.startsWith(GuiCreatorDefaults.CONFIRMATION_PREFIX)) {
            val guiConfirmEvent = GuiCancelEvent(
                event.inventory, guiCreator.identifier.replace(GuiCreatorDefaults.CONFIRMATION_PREFIX, ""),
                (event.player as Player)
            )
            Bukkit.getPluginManager().callEvent(guiConfirmEvent)
        }
    }

    @EventHandler
    fun onItemClick(event: InventoryClickEvent) {
        if (event.currentItem == null || event.currentItem!!.type == Material.AIR) return
        val itemStack = event.currentItem
        val player = event.whoClicked as Player
        val gui = GuiNavigator.getActiveGui(player) ?: return
        if (gui.inventory !== event.clickedInventory) return
        val guiItem = gui.guiCreator.correspondingGuiItem[itemStack] ?: return

        //call GuiClickEvent
        val guiClickEvent = GuiClickEvent(guiItem, event)
        Bukkit.getPluginManager().callEvent(guiClickEvent)
        if (guiClickEvent.cancelled) event.isCancelled = true
    }

    @EventHandler
    fun onGuiClick(event: GuiClickEvent) {
        val item = event.currentItem
        val view = event.view
        val player = view.player as Player
        if (item.clickSound != null) player.playSound(player.location, item.clickSound!!, 1f, 1f)
        item.clickListenerList.forEach { guiAction -> guiAction(event) }
        event.cancelled = !item.movable
    }
}